### PR TITLE
Adding spark function and tests to matrix props.

### PR DESCRIPTION
**Pull request by vprusso**
Closes: #620 

- Adding `spark` function to compute the "[spark](https://en.wikipedia.org/wiki/Spark_(mathematics))" of a matrix.
- Adding reference and doc-string containing examples.
- Adding tests for spark function.

### CODE REVIEW

File: toqito/matrix_props/spark.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    ...               [0, 1, 1, 3],
+    ...               [1, 1, 2, 5]])
+    >>> spark(A)
+    3
+
+    Notes
+    =====
+    - If all columns are linearly independent, the function returns n_cols + 1.
+    - The time complexity of this implementation is O(2^n) in the worst case,
+      where n is the number of columns.
```
Comment (purva-thakre) [inline toqito/matrix_props/spark.py:27]:
```suggestion
    - If all columns are linearly independent, the function returns $n_{\text{cols}} + 1$.
    - The time complexity of this implementation is $O(2^n)$ in the worst case,
      where $n$ is the number of columns.
```

Comment (purva-thakre) [reply toqito/matrix_props/spark.py:27]: I think you should add a linearly independent example in addition to adding this note.

File: toqito/matrix_props/spark.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    References
+    ==========
+    .. bibliography::
+        :filter: docname in docnames
+
+    :param mat: The input matrix.
+    :return: The spark of the input matrix :code:`mat`.
+
+    """
+    n_cols = mat.shape[1]
```
Comment (purva-thakre) [inline toqito/matrix_props/spark.py:38]: minor nit: You should make it clear somewhere in the docstring that this method only works for a 2D numpy array.

File: toqito/matrix_props/spark.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    for k in range(1, n_cols + 1):
+        for cols in combinations(range(n_cols), k):
+            submatrix = mat[:, cols]
+            if np.linalg.matrix_rank(submatrix) < k:
+                return k
+
```
Comment (purva-thakre) [inline toqito/matrix_props/spark.py:45]: Do you think we should add additional checks or unit tests for the case when the input matrix is not a square matrix? 

image

Screenshot taken from the linked Wikipedia page.

**Review by purva-thakre — CHANGES_REQUESTED** [review]

File: toqito/matrix_props/spark.py (hunk 1 above)
Comment (vprusso) [reply toqito/matrix_props/spark.py:29]: Fixed!

File: toqito/matrix_props/spark.py (hunk 3 above)
Comment (vprusso) [reply toqito/matrix_props/spark.py:59]: Good point. Added!

File: toqito/matrix_props/spark.py (hunk 2 above)
Comment (vprusso) [reply toqito/matrix_props/spark.py]: Done!

**Comment by vprusso** [comment]
@purva-thakre thanks for the comments! I believe I've addressed those, but let me know prior to merge!

**Review by purva-thakre — APPROVED** [review]

**Pull request merged**